src/rust/engine/grpc_util/src/tls.rs

```diff
@@ -33,71 +51,108 @@ impl TryFrom<Config> for ClientConfig {
             .map_err(|err| format!("Error creating MTLS config: {:?}", err))?
         } else {
           tls_config.with_no_client_auth()
         }
       }
       CertificateCheck::Enabled => {
-        let tls_config = match config.root_ca_certs {
-          Some(pem_bytes) => {
-            let reader = std::io::Cursor::new(pem_bytes);
-            let mut reader = std::io::BufReader::new(reader);
-            let certs = rustls_pemfile::certs(&mut reader)
-              .map_err(|err| format!("Failed to read PEM certificate: {err}"))?;
-            let mut root_cert_store = RootCertStore::empty();
-            root_cert_store.add_parsable_certificates(&certs);
-            tls_config.with_root_certificates(root_cert_store)
-          }
-          None => {
-            let native_root_certs = rustls_native_certs::load_native_certs().map_err(|err| {
-              format!(
+        let tls_config = {
+          let mut root_cert_store = RootCertStore::empty();
+
+          match config.root_ca_certs {
+            Some(certs) => {
+              for cert in &certs {
+                root_cert_store
+                  .add(cert)
+                  .map_err(|e| format!("failed adding CA cert to store: {e:?}"))?;
+              }
+            }
+            None => {
+              let native_root_certs = rustls_native_certs::load_native_certs().map_err(|err| {
+                format!(
                 "Could not discover root CA cert files to use TLS with remote caching and remote \
             execution. Consider setting `--remote-ca-certs-path` instead to explicitly point to \
             the correct PEM file.\n\n{err}",
               )
-            })?;
-            let mut root_cert_store = RootCertStore::empty();
-            for cert in native_root_certs {
-              root_cert_store.add_parsable_certificates(&[cert.0]);
+              })?;
+
+              for cert in native_root_certs {
+                root_cert_store.add_parsable_certificates(&[cert.0]);
+              }
             }
-            tls_config.with_root_certificates(root_cert_store)
           }
+
+          tls_config.with_root_certificates(root_cert_store)
         };
 
-        if let Some(MtlsConfig { key, cert_chain }) = config.mtls {
+        if let Some(MtlsConfig { cert_chain, key }) = config.mtls {
           tls_config
-            .with_client_auth_cert(
-              cert_chain_from_pem_bytes(cert_chain)?,
-              der_key_from_pem_bytes(key)?,
-            )
+            .with_client_auth_cert(cert_chain, key)
             .map_err(|err| format!("Error creating MTLS config: {:?}", err))?
         } else {
           tls_config.with_no_client_auth()
         }
       }
     };
 
-    tls_config.alpn_protocols = vec![b"h2".to_vec()];
+    // Note[TSolberg]: This causes hyper to explode -- it seems a bit odd to pre-set negotiation,
+    // but this was added for a reason presumably.
+    //tls_config.alpn_protocols = vec![b"h2".to_vec()];
```

Comment:
`tonic` sets it in its default setup logic presumably `hyper` does not. See https://github.com/hyperium/tonic/blob/6af8d5cc31a068d78c4a7937ad2eadb073e06025/examples/src/tls_rustls/server.rs#L44. I replicated that logic here since `tonic` had it.

If it is unnecessary now, then let's remove it but I would like to confirm that `hyper` does in fact set it before we do that and/or understand why `tonic` still sets it but it causes the error for you.

Reply:
We do set it here:

https://github.com/tgolsson/pants/blob/ts%2Fremote-cache-mtls/src/rust/engine/grpc_util/src/channel.rs#L51-L55
https://github.com/rustls/hyper-rustls/blob/163b3f539a497ae9c4fa65f55a8133234ef33eb3/src/connector/builder.rs#L221-L227

Reply:
But since the tlsconfig is used in many places I'm worried there's other paths that do not construct a channel in that way.

Reply:
The code at that `hyper-rustls` link sets both `h2` and `http/1.1` as "next protocols" for TLS to try as part of ALPN. gRPC only runs over `h2` though so resetting `alpn_protocols` to `h2` may be the preferred action to take since gRPC can never run over http/1.1.

Reply:
After looking again; we *might* only be getting http2. This complex builder-pattern is a bit... hard to follow. I'll have to trace out what state transitions we'd actually be going through. This is the same function but lands with only http2.

https://github.com/rustls/hyper-rustls/blob/163b3f539a497ae9c4fa65f55a8133234ef33eb3/src/connector/builder.rs#L164C1-L175C6

Reply:
Then seems like it should be safe to remove the `grpc_util` logic which sets that `alpn_protocols` then.

Reply:
Yeah. When I do a unit-test tomorrow I'll see if I can add a unit-test for that as well, after we've built the connector.